torch/autograd/gradcheck.py

```diff
@@ -250,58 +243,89 @@ def prepped_input(input: torch.Tensor, maybe_perturbed_input: Optional[torch.Ten
-            return tuple(a.clone() for a in _as_tuple(fn(*inp)))
-
+        wrapped_fn = with_prepped_inputs(fn, inputs, input_idx, x)
         input_to_perturb = x[idx]
-        nbhd_checks_fn = functools.partial(check_outputs_same_dtype_and_shape_in_neighborhood, idx=idx, eps=eps)
+        nbhd_checks_fn = functools.partial(check_outputs_same_dtype_and_shape, idx=idx, eps=eps)
+        jvp_fn = get_jvp_fn(wrapped_fn, input_to_perturb, eps, nbhd_checks_fn)
+        jacobian_cols[d_idx] = compute_numerical_jacobian_cols(jvp_fn, eps, x.is_complex(), grad_out)
+    return combine_jacobian_cols(jacobian_cols, outputs, input, input.numel())
 
-        def jvp_fn(delta):
-            return compute_numerical_gradient(wrapped_fn, input_to_perturb, delta, eps, nbhd_checks_fn)
 
-        jacobian_cols[d_idx] = compute_numerical_jacobian_cols(jvp_fn, eps, x.is_complex(), grad_out)
+def get_input_to_perturb(input):
+    if input.layout == torch._mkldnn:  # type: ignore # no attr _mkldnn
+        # Convert to dense so we can perform operations that require strided tensors
+        input_to_perturb = input.to_dense()
+    elif input.layout == torch.sparse_coo:
+        # Clone because input may require grad, and copy_ calls resize_,
+        # which is not allowed for .data
+        input_to_perturb = input.clone()
+    else:
+        input_to_perturb = input.data
+    return input_to_perturb
 
-    return combine_jacobian_cols(jacobian_cols, outputs, input, input.numel())
+
+def with_prepped_inputs(fn, inputs, input_idx, input_to_perturb, fast_mode=False):
+    def wrapped_fn():
+        inp = tuple(prepped_input(a, input_to_perturb if i == input_idx else None, fast_mode) if is_tensor_like(a) else a
+                    for i, a in enumerate(_as_tuple(inputs)))
+        return tuple(a.clone() for a in _as_tuple(fn(*inp)))
+    return wrapped_fn
+
+
+def get_jvp_fn(wrapped_fn, input_to_perturb, eps, nbhd_checks_fn):
+    def jvp_fn(delta):
+        return compute_numerical_gradient(wrapped_fn, input_to_perturb, delta, eps, nbhd_checks_fn)
+    return jvp_fn
+
+
+def get_fast_numerical_jacobian_wrt_specific_input(fn, input_idx, inputs, outputs, u, eps,
+                                                   grad_out) -> Tuple[torch.Tensor, ...]:
+    # If fast_mode=False, iter_tensor handles the below cases:
+    # basically we want to prepare the input so that it can be modified in-place and do certain
+    # operations that require the tensor to have strides
+    input = inputs[input_idx]
+    input_to_perturb = get_input_to_perturb(input)
+    wrapped_fn = with_prepped_inputs(fn, inputs, input_idx, input_to_perturb, True)
+    nbhd_checks_fn = functools.partial(check_outputs_same_dtype_and_shape, eps=eps)
+    jvp_fn = get_jvp_fn(wrapped_fn, input_to_perturb, eps, nbhd_checks_fn)
+    if u.layout != torch.sparse_coo:
+        u = u.reshape(input_to_perturb.shape)
+    return compute_numerical_jacobian_cols(jvp_fn, u * eps, input.is_complex(), grad_out)
```

Comment:
Here, in addition to factoring out the nested functions, I realized that we can actually simply return jacobian_cols instead of first allocating a bunch of zeros and then copying